src/Synth/OscilGen.cpp

```diff
@@ -493,128 +493,71 @@ class timer_t
-    int Presonance,
-    bool fillWithZeroes)
-{
-    // data
-    if(fillWithZeroes)
-    {
-        data.fillWithZeroes(); // avoid uninitialized reads
-    }
-    else
-    {
-        for(std::size_t i = 0; i < semantics.capacity(); ++i)
-        {
-            for(std::size_t j = 0; j < freqs.capacity(); ++j)
-            {
-                newrandseed(semantics[i].intVal);
-                get(data[i][j].data(), freqs[j], Presonance);
-            }
-        }
-    }
+    return std::make_pair(freqs, semantics);
 }
 
-WaveTable *OscilGen::calculateWaveTable(int Presonance) /*const*/
+wavetable_types::float32* OscilGen::calculateWaveTableData(wavetable_types::float32 freq,
+    wavetable_types::IntOrFloat semantic,
+    int Presonance)
 {
-    WaveTable* wt = new WaveTable(synth.oscilsize);
-    wt->setMode(WaveTable::WtMode::freqseed_smps);
-    std::size_t oscilsize = static_cast<std::size_t>(synth.oscilsize);
-
-    std::size_t num_semantics, num_freqs;
-
-    if(Pcurrentbasefunc == 0)
-    {
-        // sine wave - should behave the same for all freqs and semantics
-        num_semantics = num_freqs = 1;
-    }
-    else
-    {
-        num_semantics = WaveTable::num_semantics;
-        num_freqs = WaveTable::num_freqs;
-    }
-
-    Tensor1<WaveTable::float32> freqs(Shape1{num_freqs});
-    Tensor1<WaveTable::IntOrFloat> semantics(Shape1{num_semantics});
-    Tensor3<WaveTable::float32> data(
-            Shape3{num_semantics, num_freqs, oscilsize});
-
-    calculateWaveTableScales(freqs, semantics);
-    calculateWaveTableData(freqs, semantics, data, Presonance);
-
-    wt->insert(data, freqs, semantics, true);
-    return wt;
+    wavetable_types::float32* data = new wavetable_types::float32[synth.oscilsize];
+    newrandseed(semantic.intVal);
+    get(data, freq, Presonance);
+    return data;
 }
 
 WaveTable *OscilGen::allocWaveTable() const
 {
-    WaveTable* wt = new WaveTable(synth.oscilsize);
-    wt->setMode(WaveTable::WtMode::freqseed_smps);
-    std::size_t oscilsize = static_cast<std::size_t>(synth.oscilsize);
-
-    Tensor1<WaveTable::float32> freqs(Shape1{1});
-    Tensor1<WaveTable::IntOrFloat> semantics(Shape1{1});
-    Tensor3<WaveTable::float32> data(Shape3{1, 1, oscilsize});
-    wt->insert(data, freqs, semantics, true);
-    return wt;
+    return new WaveTable(static_cast<std::size_t>(synth.oscilsize));
 }
 
 void OscilGen::recalculateDefaultWaveTable(WaveTable * wt, bool fillWithZeroes) const
 {
     wt->setMode(WaveTable::WtMode::freqseed_smps);
 
-    // no allocations:
-    Tensor1<WaveTable::float32> freqs;
-    Tensor1<WaveTable::IntOrFloat> semantics;
-    Tensor3<WaveTable::float32> data;
+    // TODO: ringbuffer at size 1 will not work because the write ptr always
+    //       needs to be at least 1 behind the read pointer
+    wt->resize(Shape2{1,1});
 
-    // abuse pointer swap to steal memory from current wavetable
-    wt->insert(data, freqs, semantics, true);
+    wt->setFreq(0, 55.f);
```

Comment:
55 Hz?

Reply:
This is just the initial sine, which only computes wavetables for one frequency (no anti-aliasing required). We could use any value here, since the binary search through the wavetable frequencies with only 1 element will always return that element. 55 because the algorithm currently wavetables for 55, 110, 220, 440, ... Hz (if there's more than one frequency to calculate).

Reply:
Makes sense